data_generators.go

```diff
@@ -116,95 +149,474 @@ func NewNormalFloatGenerator(mean, stddev float64) *NormalFloatGenerator {
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	query := fmt.Sprintf("SELECT MIN(CAST(SUBSTRING_INDEX(%s, '!', 1) AS UNSIGNED)) AS min_value, MAX(CAST(SUBSTRING_INDEX(%s, '!', 1) AS UNSIGNED)) AS current_value FROM %s.%s", column, column, databaseConfig.Database, table)
+	res, err := conn.Execute(query)
+	if err != nil {
+		return nil, err
+	}
+
+	min, err := res.GetInt(0, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	current, err := res.GetInt(0, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	query = fmt.Sprintf("SELECT LENGTH(`%s`) FROM %s.%s LIMIT 1", column, databaseConfig.Database, table)
+	res, err = conn.Execute(query)
+	if err != nil {
+		return nil, err
+	}
+
+	length, err := res.GetInt(0, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UniqueStringGenerator{
+		min:     min,
+		current: atomic.NewInt64(current),
+		length:  int(length),
+	}, nil
+}
+
+func (g *UniqueStringGenerator) Generate(r *Rand) interface{} {
+	return g.GenerateTyped(r)
+}
+
+func (g *UniqueStringGenerator) GenerateTyped(r *Rand) string {
+	i := g.current.Add(1)
+	return generateUniqueStringFromInt(i, g.length)
+}
+
+func (g *UniqueStringGenerator) SampleFromExisting(r *Rand) interface{} {
+	return g.SampleFromExistingTyped(r)
+}
+
+func (g *UniqueStringGenerator) SampleFromExistingTyped(r *Rand) string {
+	max := g.current.Load()
+	// The reason we add 1 is because the values we use to generate is always the
+	// value post add rather than pre add. If we do not add 1, we will sample the
+	// first value corresponding to min and will not sample the last value
+	// corresponding to current.
+	i := r.UniformInt(g.min+1, max+1)
+	return generateUniqueStringFromInt(i, g.length)
+}
+
+type DatetimeInterval struct {
+	Start time.Time
+	End   time.Time
+}
+
+// Generates a date time value in two modes:
+//
+//  1. GenerateNow == true will cause Generate to return time.Now.
+//  2. GenerateNow == false will cause Generate to generate a random time
+//     between the intervals specified in Intervals with uniform probability
+//     distribution.
+//
+// SampleFromExisting always will sample from the Intervals. However, if
+// GenerateNow == true, then it will also sample between an extra interval
+// between when Generate() is first called and the moment the
+// SampleFromExisting call is made.
+//
+// Generate and SampleFromExisting will return a string of the time formatted
+// with YYYY-MM-DD hh:mm:ss, which is what SQL expects. GenerateTyped and
+// SampleFromExistingTyped will return time.Time.
+type UniformDatetimeGenerator struct {
+	// intervals where a datetime value will be randomly generated according to
+	// an uniform distribution. SampleFromExisting will always generate values
+	// with these intervals. Generate will use these intervals unless GenerateNow
+	// is true. In that case, Generate will add an additional interval between
+	// when the first Generate is called and time.Now.
+	intervals []DatetimeInterval
```

Comment:
One case is when you seed the data, you generate an interval. When you run the test it might be a bit later, so you need the interval on the database to form an appropriate `WHERE column > ?` query. Having multiple intervals is probably unnecessary, but I figured it's easy enough to code in here and adds some flexibility if it is needed (for example if your data is clustered on the weekends and you want to simulate multiple weekends).

Reply:
ok